runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingService.java

```diff
@@ -227,9 +235,11 @@ public void onNext(PutArtifactRequest putArtifactRequest) {
                   encodedFileName(metadata.getMetadata()), StandardResolveOptions.RESOLVE_FILE);
           LOG.debug(
               "Going to stage artifact {} to {}.", metadata.getMetadata().getName(), artifactId);
-          artifactWritableByteChannel = FileSystems.create(artifactId, MimeTypes.BINARY);
           hasher = Hashing.sha256().newHasher();
+          permittedConcurrentWrite.acquire();
```

Comment:
Since gRPC does a lot on a single thread since it multiplexes multiple RPCs on a single channel, you may be blocking other RPCs.

Reply:
I agree, this might block threads in the pool but they are anyways going to be blocked while the write is happening as we accept connections in parallel.
This take 1 call from a connection which is currently unlimited in Netty.
So I don't think this will have any added impact on the usage of grpc.

Reply:
The FileSystem may push the write to another thread unblocking the gRPC thread.